lambda/simplif.ml

```diff
@@ -119,69 +119,69 @@ let simplify_exits lam =
   | Lstringswitch(l, sw, d, _) ->
-      count l;
-      List.iter (fun (_, l) -> count l) sw;
+      count depth l;
+      List.iter (fun (_, l) -> count depth l) sw;
       begin match  d with
       | None -> ()
       | Some d -> match sw with
-        | []|[_] -> count d
-        | _ -> count d; count d (* default will get replicated *)
+        | []|[_] -> count depth d
+        | _ -> count depth d; count depth d (* default will get replicated *)
       end
-  | Lstaticraise (i,ls) -> incr_exit i 1 !try_depth; List.iter count ls
+  | Lstaticraise (i,ls) -> incr_exit i 1 try_depth; List.iter (count depth) ls
   | Lstaticcatch (l1,(i,[]),Lstaticraise (j,[])) ->
       (* i will be replaced by j in l1, so each occurrence of i in l1
          increases j's ref count *)
-      count l1 ;
+      count depth l1 ;
       let ic = get_exit i in
-      incr_exit j ic.count (max !try_depth ic.max_depth)
+      incr_exit j ic.count (max try_depth ic.max_depth)
   | Lstaticcatch(l1, (i,_), l2) ->
-      count l1;
+      count depth l1;
       (* If l1 does not contain (exit i),
          l2 will be removed, so don't count its exits *)
       if (get_exit i).count > 0 then
-        count l2
-  | Ltrywith(l1, _v, l2) -> incr try_depth; count l1; decr try_depth; count l2
-  | Lifthenelse(l1, l2, l3) -> count l1; count l2; count l3
-  | Lsequence(l1, l2) -> count l1; count l2
-  | Lwhile(l1, l2) -> count l1; count l2
-  | Lfor(_, l1, l2, _dir, l3) -> count l1; count l2; count l3
-  | Lassign(_v, l) -> count l
-  | Lsend(_k, m, o, ll, _) -> List.iter count (m::o::ll)
-  | Levent(l, _) -> count l
-  | Lifused(_v, l) -> count l
-
-  and count_default sw = match sw.sw_failaction with
+        count depth l2
+  | Ltrywith(l1, _v, l2) -> ignore (try_depth+1); count depth l1; ignore (try_depth-1); count depth l2
+  | Lifthenelse(l1, l2, l3) -> count depth l1; count depth l2; count depth l3
+  | Lsequence(l1, l2) -> count depth l1; count depth l2
+  | Lwhile(l1, l2) -> count depth l1; count depth l2
+  | Lfor(_, l1, l2, _dir, l3) -> count depth l1; count depth l2; count depth l3
+  | Lassign(_v, l) -> count depth l
+  | Lsend(_k, m, o, ll, _) -> List.iter (count depth) (m::o::ll)
+  | Levent(l, _) -> count depth l
+  | Lifused(_v, l) -> count depth l
+
+  and count_default depth sw = match sw.sw_failaction with
   | None -> ()
   | Some al ->
       let nconsts = List.length sw.sw_consts
       and nblocks = List.length sw.sw_blocks in
       if
         nconsts < sw.sw_numconsts && nblocks < sw.sw_numblocks
       then begin (* default action will occur twice in native code *)
-        count al ; count al
+        count depth al ; count depth al
       end else begin (* default action will occur once *)
         assert (nconsts < sw.sw_numconsts || nblocks < sw.sw_numblocks) ;
-        count al
+        count depth al
       end
   in
-  count lam;
-  assert(!try_depth = 0);
+  count Ident.Map.empty lam;
```

Comment:
This `Ident.Map.empty` is a very bad sign. It shows that your `depth` parameter doesn't even have the correct type.